src/collections.rs

```diff
@@ -155,3 +161,149 @@ impl From<&[u8]> for z_bytes_t {
+#[repr(C)]
+pub struct z_owned_buffer_t {
+    _inner: [usize; 5],
+}
+impl_guarded_transmute!(noderefs Option<ZBuf>, z_owned_buffer_t);
+impl Default for z_owned_buffer_t {
+    fn default() -> Self {
+        z_buffer_null()
+    }
+}
+impl From<ZBuf> for z_owned_buffer_t {
+    fn from(value: ZBuf) -> Self {
+        let value = match value.contiguous() {
+            std::borrow::Cow::Borrowed(_) => value,
+            std::borrow::Cow::Owned(value) => value.into(),
+        };
+        unsafe { core::mem::transmute(Some(value)) }
+    }
+}
+impl From<Option<ZBuf>> for z_owned_buffer_t {
+    fn from(value: Option<ZBuf>) -> Self {
+        match value {
+            Some(value) => value.into(),
+            None => z_buffer_null(),
+        }
+    }
+}
+impl core::ops::Deref for z_owned_buffer_t {
+    type Target = Option<ZBuf>;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { core::mem::transmute(self) }
+    }
+}
+impl core::ops::DerefMut for z_owned_buffer_t {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { core::mem::transmute(self) }
+    }
+}
+
+/// The gravestone value for `z_owned_buffer_t`.
+#[no_mangle]
+pub extern "C" fn z_buffer_null() -> z_owned_buffer_t {
+    unsafe { core::mem::transmute(None::<ZBuf>) }
+}
+
+/// Decrements the buffer's reference counter, destroying it if applicable.
+///
+/// `buffer` will be reset to `z_buffer_null`, preventing UB on double-frees.
+#[no_mangle]
+pub extern "C" fn z_buffer_drop(buffer: &mut z_owned_buffer_t) {
+    core::mem::drop(buffer.take())
+}
+
+/// Returns `true` if the buffer is in a valid state.
+#[no_mangle]
+pub extern "C" fn z_buffer_check(buffer: &z_owned_buffer_t) -> bool {
+    buffer.is_some()
+}
+
+/// Loans the buffer, allowing you to call functions that only need a loan of it.
+#[no_mangle]
+pub extern "C" fn z_buffer_loan(buffer: &z_owned_buffer_t) -> z_buffer_t {
+    buffer.as_ref().into()
+}
+
+/// A loan of a `z_owned_buffer_t`.
+///
+/// As it is a split buffer, it may contain more than one slice. It's number of slices is returned by `z_buffer_slice_count`.
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct z_buffer_t<'a> {
+    _inner: &'a (),
+}
+impl_guarded_transmute!(Option<&'a ZBuf>, z_buffer_t<'a>, 'a);
+impl<'a> From<z_buffer_t<'a>> for Option<&'a ZBuf> {
+    fn from(value: z_buffer_t) -> Self {
+        unsafe { core::mem::transmute(value) }
+    }
+}
+
+/// Increments the buffer's reference count, returning an owned version of the buffer.
+#[no_mangle]
+pub extern "C" fn z_buffer_clone(buffer: z_buffer_t) -> z_owned_buffer_t {
+    unsafe { Some(core::mem::transmute::<_, &ZBuf>(buffer).clone()).into() }
+}
+
+/// Returns the payload of the buffer if it is contiguous, aliasling it.
+///
+/// If the payload was not contiguous in memory, `z_bytes_null` will be returned instead.
+#[no_mangle]
+pub extern "C" fn z_buffer_payload(buffer: z_buffer_t) -> z_bytes_t {
```

Comment:
I believe the behavior of this method is counterintuitive for most users, since it might be unexpected that z_buffer_contiguous needs to be called in certain cases (which ones btw? i.e. can we reliable predict when the buffer contents is actually guaranteed to be contagious ?) to receive non-null z_bytes_t. I would rather suggest to modify the output of this method into vector of z_bytes_t, or return a struct containing a union z_bytes_t/vector of z_bytes_t to more closely resemble what we have in rust.

Reply:
According to the logic I suggest [here](https://github.com/eclipse-zenoh/zenoh-c/pull/251#discussion_r1534058102), this method won't be needed anymore